packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+ * @param leafIndex Index of the leaf node to convert to a node index.
+ * @param treeHeight Height of the binary tree.
+ * @returns the index of the leaf node in the binary tree.
+ */
+const getLeafNodeIndex = (leafIndex: number, treeHeight: number): number => {
+  return 2 ** treeHeight - 1 + leafIndex
+}
+
+/**
+ * Computes the indices of each sibling node necessary to generate a Merkle
+ * proof.
+ * @param leafIndex Index of the leaf node to get siblings for.
+ * @param treeHeight Height of the tree.
+ * @returns the indices of each sibling going up the tree.
+ */
+const getMerkleSiblingIndices = (
+  leafIndex: number,
+  treeHeight: number
+): number[] => {
+  const siblingIndices: number[] = []
+  let nodeIndex = getLeafNodeIndex(leafIndex, treeHeight)
+
+  // Go until we're at the root.
+  while (nodeIndex > 0) {
+    // Compute the sibling and add it.
+    const siblingIndex = getSiblingIndex(nodeIndex)
+    siblingIndices.push(siblingIndex)
+
+    // Go on to the parent.
+    nodeIndex = getParentIndex(nodeIndex)
+  }
+
+  return siblingIndices
+}
+
+const KEYS = {
+  BLOCKS: new BaseKey('b', ['uint32']),
+  DEPOSITS: new BaseKey('d', ['uint256']),
+  TRANSACTIONS: new BaseKey('t', ['uint32', 'uint256']),
+  TREE_HEIGHTS: new BaseKey('h', ['uint32']),
+  TREE_NODES: new BaseKey('n', ['uint32', 'uint32']),
+  LEAF_INDICES: new BaseKey('l', ['uint32', 'uint256']),
+}
+
+/**
+ * Basic ChainDB implementation that provides a nice interface to the chain
+ * database.
+ */
+export class PGChainDB implements ChainDB {
   /**
    * Creates the wrapper.
    * @param db Database to interact with.
    */
-  constructor(private db: KeyValueStore) {
-    const statePrefix = new BaseKey('s')
-    this.statedb = db.bucket(statePrefix.encode())
+  constructor(private db: KeyValueStore) {}
 
-    const historyPrefix = new BaseKey('h')
-    this.historydb = db.bucket(historyPrefix.encode())
+  /**
+   * Queries transactions in a given range.
+   * @param blockNumber Block to look in.
+   * @param start Start of the range to query.
+   * @param end End of the range to query.
+   * @returns a list of transactions in that range.
+   */
+  public async getTransactions(
+    blockNumber: number,
+    start: BigNum,
+    end: BigNum
+  ): Promise<Transaction[]> {
+    const iterator = this.db.iterator({
+      gte: KEYS.TRANSACTIONS.encode([blockNumber, bnToUint256(start)]),
+      lte: KEYS.TRANSACTIONS.encode([blockNumber, bnToUint256(end)]),
+      values: true,
+    })
+
+    const values = await iterator.values()
+    const transactions = values.map((value) => {
+      return decode(value)
+    })
+
+    return transactions
+  }
+
+  /**
+   * Queries deposits in a given range.
+   * @param start Start of the range to query.
+   * @param end End of the range to query.
+   * @returns a list of deposits for the range.
+   */
+  public async getDeposits(start: BigNum, end: BigNum): Promise<Transaction[]> {
```

Comment:
Deposits are separated from Transactions because you can't have multiple deposits on the same block. No need to store deposits on a per-block basis. Can just store them all under the same key and identified by `range.end`.